rfcs/text/0428-extending-types.md

```diff
@@ -0,0 +1,677 @@
+    "properties": {
+      "https://blockprotocol.org/@alice/property-type/birth-date": {
+        "$ref": "https://blockprotocol.org/@alice/property-type/birth-date/v/1"
+      }
+    }
+  }
+}
+```
+
+must instead use a complete schema
+
+```json
+{
+  // New, complete updateEntityType message
+  "entityTypeId": "https://blockprotocol.org/@alice/entity-type/person",
+  "schema": {
+    "$id": "https://blockprotocol.org/@alice/entity-type/person/v/2",
+    "type": "object",
+    "kind": "entityType",
+    "title": "Person",
+    "properties": {
+      "https://blockprotocol.org/@alice/property-type/name/": {
+        "$ref": "https://blockprotocol.org/@alice/property-type/name/v/1"
+      },
+      "https://blockprotocol.org/@alice/property-type/email": {
+        "$ref": "https://blockprotocol.org/@alice/property-type/email/v/1"
+      },
+      "https://blockprotocol.org/@alice/property-type/phone-number": {
+        "$ref": "https://blockprotocol.org/@alice/property-type/phone-number/v/1"
+      },
+      // The update is inlined in the existing schema
+      "https://blockprotocol.org/@alice/property-type/birth-date": {
+        "$ref": "https://blockprotocol.org/@alice/property-type/birth-date/v/1"
+      }
+    }
+  }
+}
+```
+
+given that the original schema was created as follows
+
+```json
+{
+  "$id": "https://blockprotocol.org/@alice/entity-type/person/v/1",
+  "type": "object",
+  "kind": "entityType",
+  "title": "Person",
+  "properties": {
+    "https://blockprotocol.org/@alice/property-type/name": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/name/v/1"
+    },
+    "https://blockprotocol.org/@alice/property-type/email": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/email/v/1"
+    },
+    "https://blockprotocol.org/@alice/property-type/phone-number": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/phone-number/v/1"
+    }
+  }
+}
+```
+
+The above change applies to these existing BP operations:
+
+- `updateEntityType`
+- `updatePropertyType`
+- `updateLinkType`
+
+which must all make use of complete schemas in place of partial ones. These examples originate from the [Type System RFC](./0352-graph-type-system.md#interfacing-with-types-1).
+
+While not explained in detail in the Type System RFC the `aggregateEntities` operation would need to change behavior slightly. The types returned from aggregating on a specific entity type could now also be implicitly coerced instances of entities (supertypes projected from subtype instances).
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+- The way this proposal adds type extension means that we must implement some version of property selection/projection for types, which comes with non-trivial implementation details for embedding applications.
+- This drifts further away from JSON Schema by introducing a different meaning to the `unevaluatedProperties` keyword and using it implicitly for our schemas.
```

Comment:
Is it really a different meaning? Aren't we modifying its default and changing where it may be and is inferred to be used, but not actually changing its behavior based on its presence & value?

Reply:
We are changing the meaning slightly, as we're making so that validation only occurs at the "top level type" which is not the default behavior when present